subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+	LOG_INF("data wra: %llu, %llx", SECTOR_NUM(fs->data_wra), (fs->data_wra & ADDR_OFFS_MASK));
+
+	return 0;
+}
+
+ssize_t zms_write(struct zms_fs *fs, uint32_t id, const void *data, size_t len)
+{
+	int rc, previous_sector_num = -1;
+	size_t ate_size, data_size;
+	struct zms_ate wlk_ate;
+	uint64_t wlk_addr, rd_addr;
+	uint8_t current_cycle;
+	uint32_t gc_count, required_space = 0U; /* no space, appropriate for delete ate */
+	bool prev_found = false;
+
+	if (!fs->ready) {
+		LOG_ERR("zms not initialized");
+		return -EACCES;
+	}
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+	data_size = zms_al_size(fs, len);
+
+	/* The maximum data size is sector size - 5 ate
+	 * where: 1 ate for data, 1 ate for sector close, 1 ate for empty,
+	 * 1 ate for gc done, and 1 ate to always allow a delete.
+	 */
+	if ((len > (fs->sector_size - 5 * ate_size)) || ((len > 0) && (data == NULL))) {
+		return -EINVAL;
+	}
+
+	/* find latest entry with same id */
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+	wlk_addr = fs->lookup_cache[zms_lookup_cache_pos(id)];
+
+	if (wlk_addr == ZMS_LOOKUP_CACHE_NO_ADDR) {
+		goto no_cached_entry;
+	}
+#else
+	wlk_addr = fs->ate_wra;
+#endif
+	rd_addr = wlk_addr;
+
+	while (1) {
+		rd_addr = wlk_addr;
+		rc = zms_prev_ate(fs, &wlk_addr, &wlk_ate);
+		if (rc) {
+			return rc;
+		}
+		if (wlk_ate.id == id) {
+			/* read the ate cycle only when we change the sector
+			 * or if it is the first read
+			 */
+			rc = zms_get_cycle_on_sector_change(fs, rd_addr, previous_sector_num,
+							    &current_cycle);
+			if (rc) {
+				return rc;
+			}
+			if (zms_ate_valid_different_sector(fs, &wlk_ate, current_cycle)) {
+				prev_found = true;
+				break;
+			}
+			previous_sector_num = SECTOR_NUM(rd_addr);
+		}
+
+		if (wlk_addr == fs->ate_wra) {
+			break;
+		}
+	}
+
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+no_cached_entry:
+#endif
+	if (prev_found) {
+		/* previous entry found */
+		rd_addr &= ADDR_SECT_MASK;
+		if (len > 8) {
+			rd_addr += wlk_ate.offset;
+		}
+
+		if (len == 0) {
+			/* do not try to compare with empty data */
+			if (wlk_ate.len == 0U) {
+				/* skip delete entry as it is already the
+				 * last one
+				 */
+				return 0;
+			}
+		} else if (len == wlk_ate.len) {
+			/* do not try to compare if lengths are not equal */
+			/* compare the data and if equal return 0 */
+			if (len <= 8) {
+				rc = memcmp(&wlk_ate.data, data, len);
+				if (!rc) {
+					return 0;
+				}
+			} else {
+				rc = zms_flash_block_cmp(fs, rd_addr, data, len);
+				if (rc <= 0) {
+					return rc;
+				}
+			}
+		}
+	} else {
+		/* skip delete entry for non-existing entry */
+		if (len == 0) {
+			return 0;
+		}
+	}
+
+	/* calculate required space if the entry contains data */
+	if (data_size) {
+		/* Leave space for delete ate */
+		if (len > 8) {
+			required_space = data_size + ate_size;
+		} else {
+			required_space = ate_size;
+			/* if data is in the ATE we need to make sure that we leave the ATE at
```

Comment:
@Damian-Nordic this is not the same like in NVS
By introducing the data inside ATE we can have some cases where fs->ate_wra = 0 and fs->data_wra = 0
This is the case for example when we only have values with size < 8 bytes.
fs->ate_wra will always point to the next free ATE address (in this case 0x0 which is the delete ATE)
When writing a delete ATE as the next ATE, fs->ate_wra will be decremented automatically and will overflow

Reply:
I understand that your code addresses a scenario in which:
1. There are N-1 small value ATEs written to a sector (where N is the maximum number of ATEs that fit within a sector)
2. There is a delete ATE written (in which case both `data_wra` and `ate_wra` have offset 0 within the sector).

What your code would do, it would switch to the next sector before writing the last small value ATE. But it wouldn't help for another scenario, in which there are `N-2` small value ATEs written, followed by 2 delete ATEs, right?

Therefore, my point is that it's better to ensure that a delete ATE is not written to `ate_wra` with offset component equal to 0, or even better, make sure `ate_wra` does not underflow.

Reply:
You are right @Damian-Nordic and this is also an unseen bug in NVS.
For NVS it will happen when only delete ATEs are written in a sector. The ate_wra will then underflow.
I will fix this.